source/common/http/utility.cc

```diff
@@ -57,10 +57,14 @@ bool Utility::Url::initialize(absl::string_view absolute_url) {
   // RFC allows the absolute-uri to not end in /, but the absolute path form
   // must start with
   if ((u.field_set & (1 << UF_PATH)) == (1 << UF_PATH) && u.field_data[UF_PATH].len > 0) {
-    path_ = absl::string_view(absolute_url.data() + u.field_data[UF_PATH].off,
-                              u.field_data[UF_PATH].len);
+    uint64_t path_len = u.field_data[UF_PATH].len;
+    if ((u.field_set & (1 << UF_QUERY)) == (1 << UF_QUERY) && u.field_data[UF_QUERY].len > 0) {
```

Comment:
My one remaining question - is it possible for UF_QUERY to be set and u.field_data[UF_QUERY].len == 0?

If path length is 0 we substitute a hard-coded "/".  If the +1 for length is due to the "?" being considered the delimiter I'd like to make sure that http://foo.com? would get proxied without losing the ?.  Can you add a test for that corner case?

Reply:
Ok I investigated it a bit. 
Http url parser works different for path / and ?<-query start :
e.g: http://foo.com/? it will set bit to 1 for UF_PATH but won't set UF_QUERY.
When there is query start sign(?) it goes to the next iteration and don't set UF_QUERY bit to 1. So it is impossible to get this information from parser. Same for http://foo.com?. When it gets to (?) sign and won't see another character after (?) it won't set UF_QUERY bit to 1.
I added more test cases and fixed issue when url looks like: http://foo.com?query=param

Reply:
Oh nice, thanks for digging into this, and adding the extra corner case testing!

So while this is clearly an improvement over what we have, I'm still pretty uncomfortable with /? being normalized to / unless it's illegal by the spec.

I'd be fine 
1) if this is correct as defined by some corner of the URL spec (@PiotrSikora do you know this one?)
2) landing this as is today to get most query params in this Envoy build, and leaving the bug open to sort out to retain the ? next week
3) setting the host to "everything after the beginning of the path delimiter", though we're getting close enough to merge freeze that might lose us this landing in this release.

I'd lean towards 2.  @mattklein123 thoughts/preferences?

Reply:
(2) SGTM, agreed.

Reply:
Cool.  Merged because I'd prefer the clearly-better-behavior to make next week's release.  

Lukasz, thanks so much for tackling this.  Bonus points if you'd be willing to poke through the spec or sync with @PiotrSikora on expected behavior for /? and tackle that in a follow-up :-)